akka-cluster/src/main/protobuf/ClusterMessages.proto

```diff
@@ -41,13 +41,18 @@ message Welcome {
 
 /**
  * InitJoin
- * Sends Empty
  */
+message InitJoin {
+  optional string currentConfig = 1;
+}
 
 /**
  * InitJoinAck
- * Sends an Address
  */
+message InitJoinAck {
+  required Address address = 1;
+  optional bool compatible = 2;
```

Comment:
shouldn't it send back its config also and that can be verified by the joining node also?

Reply:
I was considering that at some point. I also thought about send back a more complex type: `Unchecked` (for legacy), `Compatible` and `Incompatible`.

At the end, I convinced myself that we should not leak cluster configuration back to joining node, also not the error messages. That's why I'm only logging it on the receiving side. The joining node only sees a log saying it couldn't join and that more details are available in the cluster logging. But maybe I'm being too cautious here. 

If we run the check on both sides, then we could the following:

* joining node sends its config (stripped down);
* receiving node validates it
 * if it passes, receiving node sends its own config. Containing only the required fields.
 * if it fails, game-over

However, there is a problem. That's actually also already there.

If a node receives a `Config` instance, it can only check if it's valid using its own `JoinConfigCompatCheckers` and they may differ. The joining node may have new checkers or less checkers. Or the implementation may differ. 

But maybe that's part of the game. If a joining node has a checker for `foo.bar` and the cluster don't even have this property. The check on the joining side will fail.

Reply:
I'm not thinking of the error messages, but the second part you describe. That the joining node should also be able to validate before it actually joins. It could pass its requiredKeys in the `InitJoin` message and that would be the config sent back in the `InitJoinAck` (if first check pass).

Is your concern that it opens up a door for grabbing arbitrary config (possibly including passwords) from a cluster node by sending it `InitJoin` message? That is a valid concern but without encryption and mutual authentication there are many harmful things that can be done.

I think the most important checks are probably in the new nodes, i.e. the joining nodes, because they know about the new features or changes of defaults that enable new features. So without that part I wonder if this is too weak?

Reply:
Do we have in akka any properties that are sensitive? If we have a known set of properties we could add it, but we won't avoid rogue nodes to request properties defined by users. 

I think it's like you said, without proper security, many harmful things can happen. Maybe we should simply consider that a node joining a cluster can be trusted and that the security layer is provided on another level.

Reply:
BTW, I will add the two-side checks, ie: the joining nodes receive cluster config and use its checkers to validate it as well.

Reply:
We do something similar in Lightbend Config Checker. You can steal ideas/code from:
https://github.com/lightbend/akka-commercial-addons/blob/master/akka-diagnostics/src/main/resources/reference.conf#L239-L251
https://github.com/lightbend/akka-commercial-addons/blob/master/akka-diagnostics/src/main/scala/com/lightbend/akka/diagnostics/DiagnosticsRecorder.scala#L86